Fix #58, socket assertions in unit tests

aaaaalbert opened the pull request:
This commit addresses certain assertions and assumptions that three
unit tests had / made about the RepyV2 API. In particular,
- We cannot assume that a socket necessarily blocks and raises a
  `SocketWouldBlockError` when we use it (due to the socket buffer
  potentially swallowing/regurgitating large messages en bloc).
- Therefore, the logic to loop around sending/receiving must include
  an alternative terminal condition. (Here, we check whether we've sent
  the full message already.)
- Assertions must not depend on behavior that is not present in the
  current API implementation. (For instance, `socket.send` may report
  0 for the number of bytes sent.)
- Lastly, assertion expressions must be designed with care.
  (Obviously, checking for non-equality is not the same as checking for
  larger-than-or-equal-to.)

Also, fix a copy-pasta'd typo in these three test cases.

lukpueh commented:
For`ut_repyv2api_connectionsendwilleventuallyblock.py` and `ut_repyv2api_connectionserversendblocks.py`:

These tests would still pass if `SocketWouldBlockError` was never raised, right? Shouldn't they rather fail?  Or if not, should we maybe change the filenames and header comment?

Also, I don't really understand why a unittest that wants to provoke a full buffer does not cram in as much as possible until `SocketWouldBlockError` is raised?

Also, the header comments promise different blocking situations for `send`. I can only see one in each unittest (if we try to send although the buffer is full), and even that is not always achieved (see comment above).

Maybe I'm missing something?

aaaaalbert commented:
Thanks for your comments, @lukpueh!

> Maybe I'm missing something?

To the contrary, it looks like *I* missed something. I think I misunderstood the purpose of the `...connection...` tests by focusing on fixing the assertion (and then the byte count logic etc.), whereas the test is actually one of the "tuned" ones whose magic constant `MAXSENDSIZE` is chosen/assumed (!?!) to always completely flood the Repy socket buffer, and really should check that `SocketWouldBlockError` is raised.

I'll check against`emulcomm.py` to be sure, but I guess it will suffice to try and `send` twice the amount of `min(repy_socket_buffer_size, netsend_restriction)` to force the socket to block. I'd go with a separate restrictions file for these tests, just as with #134. (BTW, for Repy socket buffer issues see also #88 and #116.)

Also, the docstrings clearly need to explain what the purpose of the tests is!

aaaaalbert commented:
How about 145b6bb8fc888e545abb99ad39476ae61f3d3e72? Is the intent clearer now?

I'm unclear about the `assert`ions by the way. They provide a convenient way to bail out early and log at the same time, but historically we used them very sparingly.

aaaaalbert commented:
(...and here's a Travis run with a properly-sized socket buffer too, https://travis-ci.org/aaaaalbert/repy_v2/jobs/269342064 )

lukpueh commented:
Re `assert`ions:
IMHO they are more obvious than log statements plus (missing) expected output pragmas. But I guess either one is fine.

aaaaalbert commented:
Thanks again for the feedback @lukpueh, I've addressed your comments and (effectively) rebased the other work on this branch as aaaaalbert@341c62a7a8a3df7d37be379b962d82525d76b3c4.

Closing this PR, will resubmit!
